### PR TITLE
Unified `Digest::CRC#update`

### PR DESCRIPTION
**Pull request by dearblue**
The generic `Digest::CRC#update` method is provided, eliminating the need to define separate `#updates` for many CRCs.
However, the `Digest::CRC::REFLECT_INPUT` constant must be overridden by `true` or `false`.

To speed things up a bit, the following is made:

  - Use local variables instead of instance variables inside loops.
  - Suppress generation of multi-precision integers.

---

I'm not particular about the name of `REFLECT_INPUT`, so if anyone has any suggestions, I'll change it.

### CODE REVIEW

**Comment by dearblue** [comment]
For reference, I am attaching the benchmark results measured on my PC.

  - Please note that there are a number of background processes running, which may affect the results to some degree.
  - The "real" benchmark output is taken out.
  - The binary package ruby 3.2 provided by the FreeBSD project has YJIT disabled, so ruby 3.1 is used instead.
  - | `ruby32 -v` | `ruby 3.2.2 (2023-03-30 revision e51014f9c0) [amd64-freebsd13]`
    | `ruby31 -v` | `ruby 3.1.4p223 (2023-03-30 revision 957bb7cb81) [amd64-freebsd13]`
    | `jruby -v`  | `jruby 9.2.17.0 (2.5.8) 2021-03-29 84d363da97 OpenJDK 64-Bit Server VM 25.372-b07 on 1.8.0_372-b07 +jit [freebsd-x86_64]`

<table>
  <thead>
    <tr> <th> </th> <th colspan=2>ruby32</th> <th colspan=2>ruby31 --yjit</th> <th colspan=2>jruby</th> </tr>
    <tr> <th> </th> <th>main</th> <th>this PR</th> <th>main</th> <th>this PR</th> <th>main</th> <th>this PR</th> </tr>
  </thead>

  <tbody>
    <tr> <td>Digest::CRC1#update</td> <td>0.413156</td> <td>0.450639</td> <td>0.400173</td> <td>0.392028</td> <td>0.312848</td> <td>0.344441</td> </tr>
    <tr> <td>Digest::CRC5#update</td> <td>1.471989</td> <td>0.701404</td> <td>0.682974</td> <td>0.467105</td> <td>0.639566</td> <td>0.378853</td> </tr>
    <tr> <td>Digest::CRC8#update</td> <td>1.426721</td> <td>1.002237</td> <td>0.701005</td> <td>0.535166</td> <td>0.656496</td> <td>0.438962</td> </tr>
    <tr> <td>Digest::CRC8_1Wire#update</td> <td>1.435443</td> <td>0.979240</td> <td>0.785536</td> <td>0.530573</td> <td>0.658349</td> <td>0.377159</td> </tr>
    <tr> <td>Digest::CRC15#update</td> <td>1.751928</td> <td>1.721342</td> <td>0.779405</td> <td>0.666303</td> <td>0.700163</td> <td>0.700162</td> </tr>
    <tr> <td>Digest::CRC16#update</td> <td>1.402283</td> <td>1.339391</td> <td>0.733769</td> <td>0.564226</td> <td>0.710644</td> <td>0.406121</td> </tr>
    <tr> <td>Digest::CRC16CCITT#update</td> <td>1.740074</td> <td>1.661728</td> <td>0.691878</td> <td>0.707596</td> <td>0.740250</td> <td>0.789879</td> </tr>
    <tr> <td>Digest::CRC16DNP#update</td> <td>1.341073</td> <td>1.264648</td> <td>0.681094</td> <td>0.566660</td> <td>0.512899</td> <td>0.392202</td> </tr>
    <tr> <td>Digest::CRC16Genibus#update</td> <td>1.822862</td> <td>1.677790</td> <td>0.734139</td> <td>0.706260</td> <td>0.730047</td> <td>0.660704</td> </tr>
    <tr> <td>Digest::CRC16Modbus#update</td> <td>1.388758</td> <td>1.325434</td> <td>0.712974</td> <td>0.587603</td> <td>0.568764</td> <td>0.383336</td> </tr>
    <tr> <td>Digest::CRC16QT#update</td> <td>1.409243</td> <td>1.268288</td> <td>0.697429</td> <td>0.566340</td> <td>0.512600</td> <td>0.437730</td> </tr>
    <tr> <td>Digest::CRC16USB#update</td> <td>1.394310</td> <td>1.283062</td> <td>0.790167</td> <td>0.569061</td> <td>0.552077</td> <td>0.384195</td> </tr>
    <tr> <td>Digest::CRC16X25#update</td> <td>1.397066</td> <td>1.323211</td> <td>0.734347</td> <td>0.566696</td> <td>0.510791</td> <td>0.417712</td> </tr>
    <tr> <td>Digest::CRC16XModem#update</td> <td>1.776022</td> <td>1.782290</td> <td>0.731255</td> <td>0.709042</td> <td>0.720106</td> <td>0.665169</td> </tr>
    <tr> <td>Digest::CRC16ZModem#update</td> <td>1.810085</td> <td>1.664810</td> <td>0.732177</td> <td>0.696851</td> <td>0.673232</td> <td>0.607033</td> </tr>
    <tr> <td>Digest::CRC24#update</td> <td>1.783538</td> <td>1.689062</td> <td>0.730286</td> <td>0.667194</td> <td>0.760796</td> <td>0.672397</td> </tr>
    <tr> <td>Digest::CRC32#update</td> <td>1.386887</td> <td>1.287927</td> <td>0.723988</td> <td>0.628428</td> <td>0.615478</td> <td>0.432369</td> </tr>
    <tr> <td>Digest::CRC32BZip2#update</td> <td>1.761105</td> <td>1.711941</td> <td>0.737906</td> <td>0.725438</td> <td>0.660596</td> <td>0.661087</td> </tr>
    <tr> <td>Digest::CRC32c#update</td> <td>1.375790</td> <td>1.347202</td> <td>0.742967</td> <td>0.597620</td> <td>0.598534</td> <td>0.418801</td> </tr>
    <tr> <td>Digest::CRC32Jam#update</td> <td>1.377057</td> <td>1.298975</td> <td>0.689968</td> <td>0.575938</td> <td>0.654536</td> <td>0.424091</td> </tr>
    <tr> <td>Digest::CRC32MPEG#update</td> <td>1.769699</td> <td>1.696188</td> <td>0.761883</td> <td>0.665675</td> <td>0.627799</td> <td>0.660462</td> </tr>
    <tr> <td>Digest::CRC32POSIX#update</td> <td>1.728269</td> <td>1.660507</td> <td>0.759890</td> <td>0.660590</td> <td>0.690486</td> <td>0.688585</td> </tr>
    <tr> <td>Digest::CRC32XFER#update</td> <td>1.754600</td> <td>1.678388</td> <td>0.750021</td> <td>0.701695</td> <td>0.649378</td> <td>0.688124</td> </tr>
    <tr> <td>Digest::CRC64#update</td> <td>3.504403</td> <td>2.392109</td> <td>2.272752</td> <td>2.050897</td> <td>2.108015</td> <td>1.356082</td> </tr>
    <tr> <td>Digest::CRC64Jones#update</td> <td>3.461163</td> <td>2.438005</td> <td>2.322609</td> <td>2.079037</td> <td>2.145611</td> <td>1.186643</td> </tr>
    <tr> <td>Digest::CRC64XZ#update</td> <td>3.447103</td> <td>2.464268</td> <td>2.344339</td> <td>2.113575</td> <td>2.077718</td> <td>1.155891</td> </tr>
  </tbody>
</table>

File: lib/digest/crc.rb [hunk 1]
```diff
@@ -93,10 +96,52 @@ def digest_length
...
+        raise(NotImplementedError, "incompleted #{self.class} as CRC (expected REFLECT_INPUT to be not nil)")
+      end
+
+      table = @table
+      crc = @crc
+
+      if self.class::REFLECT_INPUT
+        if @width > 8
+          data.each_byte do |b|
+            crc = table[b ^ (0xff & crc)] ^ (crc >> 8)
```
Comment (dearblue) [inline lib/digest/crc.rb:118]: If we trust each element of `@table`, we can make just one more speed improvement:

```diff
-      crc = @crc
+      crc = @crc & ~(-1 << @width)

       if self.class::REFLECT_INPUT
         if @width > 8
           data.each_byte do |b|
-            crc = table[b ^ (0xff & crc)] ^ (crc >> 8)
+            crc = table[b ^ crc] ^ (crc >> 8)
```

Comment (dearblue) [reply lib/digest/crc.rb:118]: Sorry, my earlier comment is wrong.
`0xff & ... ` can be removed in three places except above.

File: lib/digest/crc.rb [hunk 2]
```diff
@@ -93,10 +96,52 @@ def digest_length
...
+        raise(NotImplementedError, "incompleted #{self.class} as CRC (expected WIDTH to be 1 or more)")
+      end
+
+      if @table.empty?
+        raise(NotImplementedError, "incompleted #{self.class} as CRC (expected TABLE to be not empty)")
+      end
+
+      if self.class::REFLECT_INPUT.nil?
+        raise(NotImplementedError, "incompleted #{self.class} as CRC (expected REFLECT_INPUT to be not nil)")
+      end
```
Comment (postmodern) [inline lib/digest/crc.rb:110]: This logic could use some specs. Just need to define a bunch of incorrect `CRC` classes in the specs:

```ruby
module TestCRC
  class CRCWithZeroWidth < Digest::CRC
    WIDTH = 0
  end
end
subject { TestCRC::CRCWithZeroWidth.new }
```

File: lib/digest/crc.rb [hunk 3]
```diff
@@ -93,10 +96,52 @@ def digest_length
...
+          end
+        else
+          data.each_byte do |b|
+            # Omit (crc >> 8) since bits upper than the lower 8 bits are always 0
+            crc = table[b ^ (0xff & crc)]
+          end
+        end
+      else
+        if @width > 8
+          drops = @width - 8
```
Comment (postmodern) [inline lib/digest/crc.rb:128]: Is there possibly a better variable name then `drops` that explains what it's doing to the CRC bits?

File: lib/digest/crc.rb [hunk 4]
```diff
@@ -93,10 +96,52 @@ def digest_length
...
+          remainmask = ~(-1 << drops)
```
Comment (postmodern) [inline lib/digest/crc.rb:129]: Should this be called `remain_mask` or `remainaing_mask`.

**Review by postmodern — CHANGES_REQUESTED** [review]
Minor changes requested. Need a `@raise` YARD tag explaining when `NotImplementedError` will be raised. Also some specs for that logic. Also some questions about variable naming.

**Comment by postmodern** [comment]
I also noticed your using `self.class::CONST`. Should we replace `self.class.const_get(:CONST)` with `self.class::CONST`?

**Comment by postmodern** [comment]
I also thought about recommending code styling changes:

* Vertically align `=`
* Add an extra newline between assignments and `data.each_byte` statement.

**Comment by dearblue** [comment]
Thank you for review. I will fix it.

File: lib/digest/crc.rb (hunk 3 above)
Comment (dearblue) [reply lib/digest/crc.rb]: Changed to `higher_bit_off`.

File: lib/digest/crc.rb (hunk 4 above)
Comment (dearblue) [reply lib/digest/crc.rb]: Changed to `remain_mask`.

File: lib/digest/crc.rb (hunk 2 above)
Comment (dearblue) [reply lib/digest/crc.rb:114]: Could I have made the changes you intended?
https://github.com/postmodern/digest-crc/pull/41/files#diff-525bf8ab75d7dddf72278222e8cbf5b2bf2f1437dbd1d368337c965d240f964bR82-R135

**Review by postmodern — APPROVED** [review]

**Pull request merged**